### ISSUE TITLE
test scenarios

### ISSUE THREAD

**Issue by cwebber** [digitalbazaar/zcapld#3]
Borrowed from [ocap-playground's toy implementation](https://gitlab.com/dustyweb/ocap-playground/blob/master/ocap-certs.rkt):

Some setup:
 - Generate public/private keypairs for the following entities: car, alyssa, ben, lem, mallet (first 4 taken from ocap-ld intro example story, last one is the malicious user)
 - Define document for car, with capability delegating/invoking keys attached to the document
 - Create document that delegates from the car to Alyssa's key (or to Alyssa, in case of entity-based).  no special caveats on this one.  The car signs off on this.  (Ideally, use a delegation *utility* to ensure that utility works)
 - Likewise for Alyssa to Ben.  On this one, add caveat which restricts the allowed action to Drive only.  Alyssa signs off on this.
 - Likewise for Ben to Lem.  (*Ideally: Add a caveat about the number of miles one can drive until.  This is a bit complicated, and it's still useful to test without adding that caveat, but it would be a nice place to attempt a very-custom caveat.  It will mean adding something about number of miles to drive to each invocation though, which does complexify the tests a little.*)

Delegation tests:

 - Invokers for the car-as-its-own-capability (the "self driving car" scenario) should be the car('s keys)
 - Invokers for car-to-alyssa-cap should be Alyssa('s key) (and car if accruing all invokers on chain)
 - Invokers for alyssa-to-ben-cap should be Ben('s key) (and Alyssa('s key) and the car('s keys))
 - Check whether the car itself is a valid capability document (no invocation, just checking whether the document itself is a "valid capability"... it should be)
 - Check whether the car-to-alyssa-cap is a valid capability document (likewise not invocation, and it should be)
 - Check whether alyssa-to-ben is a valid capability document
 - Now, have Mallet attempt to create a chain where they sign off on themselves being added to the chain (even though they shouldn't have the authority to do that). Check if the capability chain is valid (we're not invoking yet!  Just is the capability chain valid).  It shouldn't be.

Invocation tests:

 - Test an invocation of the car invoking itself with expectedTarget being the car.  Should succeed.
 - Test an invocation of Alyssa invoking the car-to-alyssa-cap against the Drive method with expectedTarget being the car; should succeed.
 - Test an invocation of Alyssa invoking the car-to-alyssa-cap against the OpenTrunk method with expectedTarget being the car; should succeed.
 - Test an invocation of Ben invoking the alyssa-to-ben-cap against the Drive method with expectedTarget being the car; should succeed.
 - Test an invocation of Ben invoking the alyssa-to-ben-cap against the OpenTrunk method with expectedTarget being the car; should fail.
 - Have Mallet generate their own attempt to delegate to themselves on the chain and try to invoke it with expectedTarget being the car.  Should fail.
 - Test an invocation of Alyssa invoking the car-to-alyssa-cap against the Drive method with expectedTarget being *something that is not the car*; should fail.
 - Test an invocation of Lem invoking the ben-to-lem-cap against the Drive method with expectedTarget being the car.  Should succeed.
 - Test an invocation of Lem invoking the ben-to-lem-cap against the OpenTrunk method with expectedTarget being the car.  Should fail.
 - BONUS: Do one with the milage, which should fail for Lem but should work fine for Ben.

Other tests to maybe add, but they might require nailing down some things in the spec:

- Revocation certificates
- Expiry
- Replay attack with reused nonce

**Comment by aljones15** [comment]
Just wanted to add the current test coverage here:

```
=============================== Coverage summary ===============================
Statements   : 83.48% ( 288/345 )
Branches     : 75.86% ( 176/232 )
Functions    : 90.32% ( 28/31 )
Lines        : 83.48% ( 288/345 )
================================================================================
ajones@vera:/work/ajones/digital_bazaar/ocapld.js$ vim package.json 
ajones@vera:/work/ajones/digital_bazaar/ocapld.js$ npm run coverage-report

> ocapld@1.7.1-0 coverage-report /work/ajones/digital_bazaar/ocapld.js
> nyc report

-------------------------|----------|----------|----------|----------|-------------------|
File                     |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
-------------------------|----------|----------|----------|----------|-------------------|
All files                |    83.48 |    75.86 |    90.32 |    83.48 |                   |
 CapabilityDelegation.js |    90.91 |    80.77 |      100 |    90.91 | 59,72,114,131,156 |
 CapabilityInvocation.js |    88.24 |    77.78 |      100 |    88.24 |... 78,106,115,135 |
 Caveat.js               |       50 |    33.33 |    33.33 |       50 |          14,28,39 |
 ExpirationCaveat.js     |    94.74 |     87.5 |      100 |    94.74 |                19 |
 constants.js            |      100 |      100 |      100 |      100 |                   |
 index.js                |      100 |      100 |      100 |      100 |                   |
 utils.js                |    79.61 |       75 |    94.12 |    79.61 |... 79,490,675,690 |
-------------------------|----------|----------|----------|----------|-------------------|
```

Over all this library is really well tested however `CapabilityInvocation` is is lacking multiple lines that are not covered.

Missing tests:

`CapabilityDelegation.js`
  - constructor caveat is not an array
  - validate document does not contain a `parentCapability`
  - document.parentCapability !== verifiedParentCapability.id
  - if validateResult is not valid should THROW
  - capabilityChain is not an array

`CapabilityInvocation.js`
  - proof has no capability
  -  if(!utils.isInvoker({capability, verificationMethod})) {
  - SHOULD throw !validateResult.valid
  - capabilityAction && typeof capabilityAction !== 'string'
  - more that we don't know about yet

`Caveat.js` is also missing tests etc.

**Comment by dlongley** [comment]
Closing as either these will not be implemented or there are now tests in #73 that address many of the cases.
